hbase-server/src/main/java/org/apache/hadoop/hbase/io/hfile/bucket/BucketCache.java

```diff
@@ -642,12 +643,14 @@ void blockEvicted(BlockCacheKey cacheKey, BucketEntry bucketEntry, boolean decre
     blocksByHFile.remove(cacheKey);
     if (decrementBlockNumber) {
       this.blockNumber.decrement();
+      if (ioEngine.isPersistent()) {
+        removeFileFromPrefetch(cacheKey.getHfileName());
+      }
     }
     if (evictedByEvictionProcess) {
       cacheStats.evicted(bucketEntry.getCachedTime(), cacheKey.isPrimary());
```

Comment:
So we could throw exceptions here?

Reply:
What about this one? Why we need to move this up?

Reply:
Ping @wchevreuil

Reply:
Sorry, missed the previous comment. 

It was wrong to call the removeFileFromPrefetch if we are not decrementing the block number. In some occasions, we do replace the same block and call blockEvicted with decrementBlockNumber set to false. This means even though we evicted, we still have that block for the file and we shouldn't remove it from the list of fully cached files.